<a href="https://colab.research.google.com/github/mertcolab/Advent2020/blob/main/p4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Day 4: Passport Processing

### Part 1 

You arrive at the airport only to realize that you grabbed your North Pole Credentials instead of your passport. While these documents are extremely similar, North Pole Credentials aren't issued by a country and therefore aren't actually valid documentation for travel in most of the world.


It seems like you're not the only one having problems, though; a very long line has formed for the automatic passport scanners, and the delay could upset your travel itinerary.


Due to some questionable network security, you realize you might be able to solve both of these problems at the same time.


The automatic passport scanners are slow because they're having trouble detecting which passports have all required fields. The expected fields are as follows:


* byr (Birth Year)
* iyr (Issue Year)
* eyr (Expiration Year)
* hgt (Height)
* hcl (Hair Color)
* ecl (Eye Color)
* pid (Passport ID)
* cid (Country ID)

Passport data is validated in batch files (your puzzle input). Each passport is represented as a sequence of key:value pairs separated by spaces or newlines. Passports are separated by blank lines.


Here is an example batch file containing four passports:


* ecl:gry pid:860033327 eyr:2020 hcl:#fffffd <br>
  byr:1937 iyr:2017 cid:147 hgt:183cm


* iyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884<br>
hcl:#cfa07d byr:1929

* hcl:#ae17e1 iyr:2013<br>
eyr:2024<br>
ecl:brn pid:760753108 byr:1931<br>
hgt:179cm

* hcl:#cfa07d eyr:2025 pid:166559648<br>
iyr:2011 ecl:brn hgt:59in<br>

The first passport is valid - all eight fields are present. 

The second passport  is invalid - it is missing hgt (the Height field).


The third passport is interesting; the only missing field is cid, so it looks like data from North Pole Credentials, not a passport at all! Surely, nobody would mind if you made the system temporarily ignore missing cid fields. Treat this "passport" as valid.


The fourth passport is missing two fields, cid and byr. Missing cid is fine, but missing any other field is not, so this passport is invalid.


According to the above rules, your improved system would report 2 valid passports.


Count the number of valid passports - those that have all required fields. Treat cid as optional. In your batch file, how many passports are valid?




In [1]:
from google.colab import drive
drive.mount('gdrive')

Mounted at gdrive


In [2]:
path = "/content/gdrive/My Drive/Colab Notebooks/Advent2020/p4_input.txt"
myfile = open(path)

In [3]:
import re

In [4]:
data = []

with open(path) as file:
  record = []
  for line in file.read().split('\n'):
    if line != '':
      for element in line.split(' '):
        record.append(element)
    else:
      data.append(record)
      record = []

print(len(data))

279


In [5]:
counter = 0
for i in range(len(data)):
    entry = {'pid':0, 'byr':0, 'iyr':0, 'eyr':0, 'hgt':0, 'hcl':0, 'ecl':0 }
    checklist =  {'byr':1, 'iyr':1, 'eyr':1, 'hgt':1, 'hcl':1, 'ecl':1, 'pid':1}
    for member in data[i]:
      key, value = member.split(':')
      
      # ignoring cid
      if key!='cid':
        checklist[key]-=1
        entry[key]=value
    
    # all values of checklist is exactly zero and there are exactly 7 keys.
    if (all(value == 0 for value in checklist.values()) and len(checklist.keys())==7):
      counter+=1
  

print("There are {} valid entries".format(counter))

There are 213 valid entries


### Part 2

The line is moving more quickly now, but you overhear airport security talking about how passports with invalid data are getting through. Better add some data validation, quick!

You can continue to ignore the cid field, but each other field has strict rules about what values are valid for automatic validation:

byr (Birth Year) - four digits; at least 1920 and at most 2002.<br>
iyr (Issue Year) - four digits; at least 2010 and at most 2020.<br>
eyr (Expiration Year) - four digits; at least 2020 and at most 2030.<br>
hgt (Height) - a number followed by either cm or in:<br>
If cm, the number must be at least 150 and at most 193.<br>
If in, the number must be at least 59 and at most 76.<br>
hcl (Hair Color) - a # followed by exactly six characters 0-9 or a-f.<br>
ecl (Eye Color) - exactly one of: amb blu brn gry grn hzl oth.<br>
pid (Passport ID) - a nine-digit number, including leading zeroes.<br>
cid (Country ID) - ignored, missing or not.<br>
Your job is to count the passports where all required fields are both present and valid according to the above rules. Here are some example values:<br>

```
byr valid:   2002
byr invalid: 2003
```
```
hgt valid:   60in
hgt valid:   190cm
hgt invalid: 190in
hgt invalid: 190
```
```
hcl valid:   #123abc
hcl invalid: #123abz
hcl invalid: 123abc
```
```
ecl valid:   brn
ecl invalid: wat
```
```
pid valid:   000000001
pid invalid: 0123456789
```
Here are some invalid passports:
```
eyr:1972 cid:100
hcl:#18171d ecl:amb hgt:170 pid:186cm iyr:2018 byr:1926
```

```
iyr:2019
hcl:#602927 eyr:1967 hgt:170cm
ecl:grn pid:012533040 byr:1946
```
```
hcl:dab227 iyr:2012
ecl:brn hgt:182cm pid:021572410 eyr:2020 byr:1992 cid:277
```
```
hgt:59cm ecl:zzz
eyr:2038 hcl:74454a iyr:2023
pid:3556412378 byr:2007
```
Here are some valid passports:

```
pid:087499704 hgt:74in ecl:grn iyr:2012 eyr:2030 byr:1980
hcl:#623a2f
```
```
eyr:2029 ecl:blu cid:129 byr:1989
iyr:2014 pid:896056539 hcl:#a97842 hgt:165cm
```
```
hcl:#888785
hgt:164cm byr:2001 iyr:2015 cid:88
pid:545766238 ecl:hzl
eyr:2022
```
```
iyr:2010 hgt:158cm hcl:#b6652a ecl:blu byr:1944 eyr:2021 pid:093154719
```
Count the number of valid passports - those that have all required fields and valid values. Continue to treat cid as optional. In your batch file, how many passports are valid?

#### Supplementary Function Definitions

In [6]:

def regex_check(rule, input):
  return re.match(rule, input ) is not None

eyecolorlist = ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']

def year_limit_check(mincon, maxcon, value):
  value = int(value)
  return mincon <= value  and value<= maxcon

def is_in_list(list, value):
  for i in list:
    if i == value:
      return True
  return False

In [8]:
counter = 0
listofvalid2 = []
for i in range(len(data)):
  entry = {'pid':0, 'byr':0, 'iyr':0, 'eyr':0, 'hgt':0, 'hcl':0, 'ecl':0 }
  checklist =  {'byr':1, 'iyr':1, 'eyr':1, 'hgt':1, 'hcl':1, 'ecl':1, 'pid':1}
  for member in data[i]:
    key, value = member.split(':')
    
    # ignoring cid 
    if key!='cid':
      checklist[key]-=1
      entry[key]=value

  # all values of checklist is exactly zero and there are exactly 7 keys. 
  if (all(value == 0 for value in checklist.values()) and len(checklist.keys())==7):
    # year limits 
    if (year_limit_check(1920, 2002, entry['byr']) and year_limit_check(2010, 2020, entry['iyr']) and year_limit_check(2020, 2030, entry['eyr']) ):
      # pid rule check 9 digit number
      if (regex_check('[0-9]{9}', entry['pid'])):
        #color code check hair # 6 hex digit; eye color 3 letters and (height in inches or cm)
        if (regex_check('#[a-f0-9]{6}', entry['hcl']) and is_in_list(eyecolorlist, entry['ecl']) and ( (entry['hgt'][-2:]=='in' and 59<=int(entry['hgt'][:-2]) and 76>=int(entry['hgt'][:-2])) or (entry['hgt'][-2:]=='cm' and 150<=int(entry['hgt'][:-2]) and 193>=int(entry['hgt'][:-2])) )): 
          counter+=1
          listofvalid2.append(entry)
          
print("There are {} valid entries".format(counter))

There are 148 valid entries


My initial version was giving 147, now it gives 148. There should be a mistake but I won't go into details for now.

